x/ibc/03-connection/keeper/handshake.go

```diff
@@ -0,0 +1,242 @@
+func (k Keeper) ConnOpenInit(
+	ctx sdk.Context,
+	connectionID, // identifier
+	clientID string,
+	counterparty types.Counterparty, // desiredCounterpartyConnectionIdentifier, counterpartyPrefix, counterpartyClientIdentifier
+) error {
+	// TODO: validateConnectionIdentifier(identifier)
+	_, found := k.GetConnection(ctx, connectionID)
+	if found {
+		return sdkerrors.Wrap(types.ErrConnectionExists(k.codespace), "cannot initialize connection")
+	}
+
+	// connection defines chain A's ConnectionEnd
+	connection := types.NewConnectionEnd(clientID, counterparty, k.getCompatibleVersions())
+	connection.State = types.INIT
+
+	k.SetConnection(ctx, connectionID, connection)
+	err := k.addConnectionToClient(ctx, clientID, connectionID)
+	if err != nil {
+		sdkerrors.Wrap(err, "cannot initialize connection")
+	}
+
+	k.Logger(ctx).Info(fmt.Sprintf("connection %s state updated: NONE -> INIT", connectionID))
+	return nil
+}
+
+// ConnOpenTry relays notice of a connection attempt on chain A to chain B (this
+// code is executed on chain B).
+//
+// NOTE: here chain A acts as the counterparty
+func (k Keeper) ConnOpenTry(
+	ctx sdk.Context,
+	connectionID string, // desiredIdentifier
+	counterparty types.Counterparty, // counterpartyConnectionIdentifier, counterpartyPrefix and counterpartyClientIdentifier
+	clientID string,
+	counterpartyVersions []string,
+	proofInit ics23.Proof,
+	proofHeight uint64,
+	consensusHeight uint64,
+) error {
+	// TODO: validateConnectionIdentifier(identifier)
+	if consensusHeight > uint64(ctx.BlockHeight()) {
+		return errors.New("invalid consensus height") // TODO: sdk.Error
+	}
+
+	expectedConsensusState, found := k.clientKeeper.GetConsensusState(ctx, clientID)
+	if !found {
+		return errors.New("client consensus state not found") // TODO: use ICS02 error
+	}
+
+	// expectedConn defines Chain A's ConnectionEnd
+	// NOTE: chain A's counterparty is chain B (i.e where this code is executed)
+	// TODO: prefix should be `getCommitmentPrefix()`
+	expectedCounterparty := types.NewCounterparty(counterparty.ClientID, connectionID, counterparty.Prefix)
+	expectedConn := types.NewConnectionEnd(clientID, expectedCounterparty, counterpartyVersions)
+	expectedConn.State = types.INIT
+
+	// chain B picks a version from Chain A's available versions
+	version := k.pickVersion(counterpartyVersions)
+
+	// connection defines chain B's ConnectionEnd
+	connection := types.NewConnectionEnd(clientID, counterparty, []string{version})
+	expConnBz, err := k.cdc.MarshalBinaryLengthPrefixed(expectedConn)
+	if err != nil {
+		return err
+	}
+
+	ok := k.VerifyMembership(
+		ctx, connection, proofHeight, proofInit,
+		types.ConnectionPath(connectionID), expConnBz,
+	)
+	if !ok {
+		return errors.New("couldn't verify connection membership on counterparty's client") // TODO: sdk.Error
+	}
+
+	expConsStateBz, err := k.cdc.MarshalBinaryLengthPrefixed(expectedConsensusState)
+	if err != nil {
+		return err
+	}
+
+	ok = k.VerifyMembership(
```

Comment:
I would consider abstracting this out (the fetch of consensus state and the verification)

Reply:
on a separate private function you mean?